File: backend/app/api/memory.py
from __future__ import annotations

from typing import Annotated, Literal, Protocol

from fastapi import APIRouter, Depends, Query

from app.api.dependencies import get_workflow_memory_policy_service
from app.schemas.workflow_memory import (
    WORKFLOW_MEMORY_REVIEW_LIST_DEFAULT_LIMIT,
    WORKFLOW_MEMORY_REVIEW_LIST_MAX_LIMIT,
    WorkflowMemoryAuditEventListRead,
    WorkflowMemoryPolicyStatus,
    WorkflowMemoryProposalListRead,
    WorkflowMemoryQuarantineListRead,
    WorkflowMemoryReviewActionRead,
    WorkflowMemoryReviewActionRequest,
)

router = APIRouter(prefix="/memory", tags=["memory"])


class WorkflowMemoryReviewService(Protocol):
    def list_review_proposals(
        self,
        *,
        status: WorkflowMemoryPolicyStatus | None,
        limit: int,
        offset: int,
    ) -> WorkflowMemoryProposalListRead: ...

    def approve_review_pending_proposal(
        self,
        *,
        proposal_id: str,
        reason: str | None,
    ) -> WorkflowMemoryReviewActionRead: ...

    def reject_review_pending_proposal(
        self,
        *,
        proposal_id: str,
        reason: str | None,
    ) -> WorkflowMemoryReviewActionRead: ...

    def list_audit_events(
        self,
        *,
        limit: int,
        offset: int,
    ) -> WorkflowMemoryAuditEventListRead: ...

    def list_quarantine(
        self,
        *,
        unresolved_only: bool,
        limit: int,
        offset: int,
    ) -> WorkflowMemoryQuarantineListRead: ...


@router.get("/proposals", response_model=WorkflowMemoryProposalListRead)
def list_memory_proposals(
    service: Annotated[WorkflowMemoryReviewService, Depends(get_workflow_memory_policy_service)],
    status: Annotated[WorkflowMemoryPolicyStatus | Literal["all"], Query()] = (
        WorkflowMemoryPolicyStatus.REVIEW_PENDING
    ),
    limit: Annotated[int, Query(ge=1, le=WORKFLOW_MEMORY_REVIEW_LIST_MAX_LIMIT)] = (
        WORKFLOW_MEMORY_REVIEW_LIST_DEFAULT_LIMIT
    ),
    offset: Annotated[int, Query(ge=0)] = 0,
) -> WorkflowMemoryProposalListRead:
    return service.list_review_proposals(
        status=None if status == "all" else status,
        limit=limit,
        offset=offset,
    )


@router.post(
    "/proposals/{proposal_id}/actions/approve",
    response_model=WorkflowMemoryReviewActionRead,
)
def approve_memory_proposal(
    proposal_id: str,
    payload: WorkflowMemoryReviewActionRequest,
    service: Annotated[WorkflowMemoryReviewService, Depends(get_workflow_memory_policy_service)],
) -> WorkflowMemoryReviewActionRead:
    return service.approve_review_pending_proposal(
        proposal_id=proposal_id,
        reason=payload.reason,
    )


@router.post(
    "/proposals/{proposal_id}/actions/reject",
    response_model=WorkflowMemoryReviewActionRead,
)
def reject_memory_proposal(
    proposal_id: str,
    payload: WorkflowMemoryReviewActionRequest,
    service: Annotated[WorkflowMemoryReviewService, Depends(get_workflow_memory_policy_service)],
) -> WorkflowMemoryReviewActionRead:
    return service.reject_review_pending_proposal(
        proposal_id=proposal_id,
        reason=payload.reason,
    )


@router.get("/audit-events", response_model=WorkflowMemoryAuditEventListRead)
def list_memory_audit_events(
    service: Annotated[WorkflowMemoryReviewService, Depends(get_workflow_memory_policy_service)],
    limit: Annotated[int, Query(ge=1, le=WORKFLOW_MEMORY_REVIEW_LIST_MAX_LIMIT)] = (
        WORKFLOW_MEMORY_REVIEW_LIST_DEFAULT_LIMIT
    ),
    offset: Annotated[int, Query(ge=0)] = 0,
) -> WorkflowMemoryAuditEventListRead:
    return service.list_audit_events(limit=limit, offset=offset)


@router.get("/quarantine", response_model=WorkflowMemoryQuarantineListRead)
def list_memory_quarantine(
    service: Annotated[WorkflowMemoryReviewService, Depends(get_workflow_memory_policy_service)],
    unresolved_only: Annotated[bool, Query(alias="unresolvedOnly")] = True,
    limit: Annotated[int, Query(ge=1, le=WORKFLOW_MEMORY_REVIEW_LIST_MAX_LIMIT)] = (
        WORKFLOW_MEMORY_REVIEW_LIST_DEFAULT_LIMIT
    ),
    offset: Annotated[int, Query(ge=0)] = 0,
) -> WorkflowMemoryQuarantineListRead:
    return service.list_quarantine(
        unresolved_only=unresolved_only,
        limit=limit,
        offset=offset,
    )

File: frontend/src/lib/api/memory.ts
import { requestPlatform, toPathSegment, toQueryRecord } from "../api-client";
import type {
  WorkflowMemoryAuditEventListRead,
  WorkflowMemoryListParams,
  WorkflowMemoryProposalListParams,
  WorkflowMemoryProposalListRead,
  WorkflowMemoryQuarantineListParams,
  WorkflowMemoryQuarantineListRead,
  WorkflowMemoryReviewActionRead,
  WorkflowMemoryReviewActionRequest,
} from "../types/memory";

function normalizeListParams<TParams extends WorkflowMemoryListParams>(
  params: TParams = {} as TParams,
): TParams {
  return {
    ...params,
    limit: params.limit ?? 50,
    offset: params.offset ?? 0,
  };
}

export function normalizeWorkflowMemoryProposalParams(
  params: WorkflowMemoryProposalListParams = {},
): WorkflowMemoryProposalListParams {
  return {
    ...normalizeListParams(params),
    status: params.status ?? "review_pending",
  };
}

export function normalizeWorkflowMemoryAuditParams(
  params: WorkflowMemoryListParams = {},
): WorkflowMemoryListParams {
  return normalizeListParams(params);
}

export function normalizeWorkflowMemoryQuarantineParams(
  params: WorkflowMemoryQuarantineListParams = {},
): WorkflowMemoryQuarantineListParams {
  return {
    ...normalizeListParams(params),
    unresolvedOnly: params.unresolvedOnly ?? true,
  };
}

export function listWorkflowMemoryProposals(
  params: WorkflowMemoryProposalListParams = {},
  signal?: AbortSignal,
): Promise<WorkflowMemoryProposalListRead> {
  return requestPlatform<WorkflowMemoryProposalListRead>("/memory/proposals", {
    query: toQueryRecord(normalizeWorkflowMemoryProposalParams(params)),
    signal,
  });
}

export function approveWorkflowMemoryProposal(
  proposalId: string,
  payload: WorkflowMemoryReviewActionRequest = {},
  signal?: AbortSignal,
): Promise<WorkflowMemoryReviewActionRead> {
  return requestPlatform<WorkflowMemoryReviewActionRead>(
    `/memory/proposals/${toPathSegment(proposalId)}/actions/approve`,
    { body: payload, method: "POST", signal },
  );
}

export function rejectWorkflowMemoryProposal(
  proposalId: string,
  payload: WorkflowMemoryReviewActionRequest = {},
  signal?: AbortSignal,
): Promise<WorkflowMemoryReviewActionRead> {
  return requestPlatform<WorkflowMemoryReviewActionRead>(
    `/memory/proposals/${toPathSegment(proposalId)}/actions/reject`,
    { body: payload, method: "POST", signal },
  );
}

export function listWorkflowMemoryAuditEvents(
  params: WorkflowMemoryListParams = {},
  signal?: AbortSignal,
): Promise<WorkflowMemoryAuditEventListRead> {
  return requestPlatform<WorkflowMemoryAuditEventListRead>(
    "/memory/audit-events",
    { query: toQueryRecord(normalizeWorkflowMemoryAuditParams(params)), signal },
  );
}

export function listWorkflowMemoryQuarantine(
  params: WorkflowMemoryQuarantineListParams = {},
  signal?: AbortSignal,
): Promise<WorkflowMemoryQuarantineListRead> {
  return requestPlatform<WorkflowMemoryQuarantineListRead>("/memory/quarantine", {
    query: toQueryRecord(normalizeWorkflowMemoryQuarantineParams(params)),
    signal,
  });
}

export const memoryApi = {
  approveProposal: approveWorkflowMemoryProposal,
  auditEvents: listWorkflowMemoryAuditEvents,
  proposals: listWorkflowMemoryProposals,
  quarantine: listWorkflowMemoryQuarantine,
  rejectProposal: rejectWorkflowMemoryProposal,
} as const;
